relay-server/src/actors/envelopes.rs

```diff
@@ -46,6 +46,51 @@ pub enum QueueEnvelopeError {
     TooManyEnvelopes,
 }
 
+/// TODO(ja): Doc
+#[derive(Debug)]
+pub struct BufferGuard {
+    inner: Semaphore,
+    total: usize,
+}
+
+impl BufferGuard {
+    /// TODO(ja): Doc
+    pub fn new(config: &Config) -> Self {
+        let total = config.envelope_buffer_size();
+        let inner = Semaphore::new(total);
+        Self { inner, total }
+    }
+
+    /// TODO(ja): Doc
+    pub fn capacity(&self) -> usize {
+        self.inner.capacity()
+    }
+
+    /// TODO(ja): Doc
+    pub fn utilization(&self) -> usize {
+        self.total.saturating_sub(self.capacity())
+    }
+
+    /// TODO(ja): Doc
+    pub fn enter(&self, envelope: &Envelope) -> Result<EnvelopeContext, QueueEnvelopeError> {
```

Comment:
`BufferGuard` could theoretically also be written as a simple async service that wraps the semaphore and responds with the EnvelopeContext. However, its behavior is much more like a higher-level semaphore, so I chose to keep it just that. That also avoids any overheads of sending messages in and out.